app/server/src/db/sessionStore.ts

```diff
@@ -40,28 +41,77 @@ export class SessionStore {
 
     private sessionKey = (name: string) => `${this._sessionPrefix}${name}`;
 
+    private static hashForData = (data: string) => md5(data);
+
+    private saveMissingSessionHash = async (sessionId: string, hash: string) => {
+        await this._redis.hset(this.sessionKey("hash"), sessionId, hash);
+    };
+
     async saveSession(id: string, data: string) {
+        const hash = SessionStore.hashForData(data);
         await this._redis.pipeline()
             .hset(this.sessionKey("time"), id, new Date(Date.now()).toISOString())
             .hset(this.sessionKey("data"), id, data)
+            .hset(this.sessionKey("hash"), id, hash)
             .exec();
     }
 
-    async getSessionsMetadata(ids: string[]) {
+    async getSessionsMetadata(ids: string[], removeDuplicates: boolean = true) {
         return Promise.all([
             this._redis.hmget(this.sessionKey("time"), ...ids),
             this._redis.hmget(this.sessionKey("label"), ...ids),
-            this._redis.hmget(this.sessionKey("friendly"), ...ids)
-        ]).then((values) => {
+            this._redis.hmget(this.sessionKey("friendly"), ...ids),
+            this._redis.hmget(this.sessionKey("hash"), ...ids)
+        ]).then(async (values) => {
             const times = values[0];
             const labels = values[1];
             const friendlies = values[2];
-            const allResults = ids.map((id: string, idx: number) => {
-                return {
-                    id, time: times[idx], label: labels[idx], friendlyId: friendlies[idx]
-                };
+            const hashes = values[3];
+
+            const buildSessionMetadata = (id: string, idx: number) => ({
+                id, time: times[idx], label: labels[idx], friendlyId: friendlies[idx]
             });
-            return allResults.filter((session) => session.time !== null) as SessionMetadata[];
+
+            let allResults;
+            if (removeDuplicates) {
+                // We return all labelled sessions, and also the latest session for any hash (labelled or unlabelled)
+                const labelledSessions: SessionMetadata[] = [];
+                const latestByHash: Record<string, SessionMetadata> = {};
+
+                const saveMissingHashes = [];
+                // eslint-disable-next-line no-restricted-syntax
+                for await (const [idx, id] of ids.entries()) {
```

Comment:
eslint doesn't like plain old `for`s but this is preferable for including async in the loop